### PR TITLE
test: leader package

### PR DESCRIPTION
**Pull request by matt2e**

### CODE REVIEW

File: backend/controller/leader/leader_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	"time"
+
+	"github.com/TBD54566975/ftl/backend/controller/leases"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/alecthomas/assert/v2"
+	"github.com/alecthomas/types/optional"
+)
+
+func TestExistingLeaseForURL(t *testing.T) {
+	// Test that if the a lease exists with the current URL, neither a leader or follower is created
```
Comment (alecthomas) [inline backend/controller/leader/leader_test.go:18]: Buy "current" URL do you mean our own URL or something else?

File: backend/controller/leases/fake_lease.go [hunk 2]
```diff
@@ -11,32 +12,51 @@ import (
...
+		cancelCtx()
+		return nil, nil, ErrConflict
+	}
+	return newLease, leaseCtx, nil
 }
 
 func (f *FakeLeaser) GetLeaseInfo(ctx context.Context, key Key, metadata any) (expiry time.Time, err error) {
-	if _, ok := f.leases.Load(key.String()); ok {
-		return time.Now().Add(time.Minute), nil
+	if lease, ok := f.leases.Load(key.String()); ok {
```
Comment (alecthomas) [inline backend/controller/leases/fake_lease.go:43]: Invert logic and early exit?

File: backend/controller/leases/fake_lease.go [hunk 3]
```diff
@@ -11,32 +12,51 @@ import (
...
+		if md, ok := lease.metadata.Get(); ok && metadata != nil {
```
Comment (alecthomas) [inline backend/controller/leases/fake_lease.go:44]: Invert logic and early exit?

**Review by alecthomas — APPROVED** [review]

File: backend/controller/leader/leader_test.go (hunk 1 above)
Comment (matt2e) [reply backend/controller/leader/leader_test.go:18]: The lease has metadata set to the url of the leader. This is testing whether the coordinator's url matches the leader's url, but the coordinator does not have the lease, which can happen if a lease loses the connection, or a controller starts up at a url that was used by a controller that died.
I'll reword it, "current" isn't right.

**Pull request merged**